programs/sunrise-stake/src/sunrise_spl/instructions/withdraw_stake.rs

```diff
@@ -0,0 +1,158 @@
+///  userdata: amount of pool tokens to withdraw
+
+#[derive(Accounts)]
+pub struct SplWithdrawStake<'info> {
+    #[account(has_one = gsol_mint)]
+    pub state: Box<Account<'info, State>>,
+    pub gsol_mint: Box<Account<'info, Mint>>,
+    #[account(
+        seeds = [state.key().as_ref(), seeds::GSOL_MINT_AUTHORITY],
+        bump = state.gsol_mint_authority_bump
+    )]
+    pub gsol_mint_authority: SystemAccount<'info>,
+
+    pub user: Signer<'info>,
+    #[account(
+        mut,
+        token::mint = gsol_mint,
+        token::authority = user,
+    )]
+    pub user_gsol_token_account: Account<'info, TokenAccount>,
+    #[account(mut)]
+    /// CHECK:
+    pub user_new_stake_account: AccountInfo<'info>, //Should be an uninitialized stake account
+
+    #[account(
+        mut, has_one = state, has_one = pool_token_vault,
+        seeds = [b"pool".as_ref(), stake_pool.key().as_ref()], bump
+    )]
+    pub pool_details: Account<'info, SplPoolDetails>,
+    #[account(mut)]
+    pub pool_token_vault: Account<'info, TokenAccount>,
+
+    #[account(mut)]
+    /// CHECK:
+    pub stake_pool: AccountInfo<'info>,
+    /// CHECK:
+    pub stake_pool_withdraw_authority: AccountInfo<'info>,
+    #[account(mut)]
+    /// CHECK:
+    pub validator_stake_list: AccountInfo<'info>,
+    #[account(mut)]
+    /// CHECK:
+    pub stake_account_to_split: AccountInfo<'info>, //either validator/reserve
+    #[account(mut)]
+    /// CHECK:
+    pub manager_fee_account: AccountInfo<'info>,
+    #[account(mut)]
+    /// CHECK:
+    pub stake_pool_token_mint: AccountInfo<'info>,
+    /// CHECK:
+    pub sysvar_clock: AccountInfo<'info>,
+    /// CHECK:
+    pub stake_pool_program: AccountInfo<'info>,
+    /// CHECK:
+    pub native_stake_program: AccountInfo<'info>,
+
+    pub token_program: Program<'info, Token>,
+}
+
+impl<'info> SplWithdrawStake<'info> {
+    fn check_stake_pool_program(&self) -> Result<()> {
+        require_keys_eq!(*self.stake_pool_program.key, spl_stake_pool::ID);
+        Ok(())
+    }
+
+    pub fn withdraw_stake(&self, amount: u64) -> Result<()> {
+        self.check_stake_pool_program()?;
+
+        let bump = self.pool_details.bump;
+        let signer_seeds = &[b"pool".as_ref(), &self.stake_pool.key.as_ref(), &[bump]];
+        let signer_seeds = &[&signer_seeds[..]];
+
+        invoke_signed(
+            &spl_stake_pool::instruction::withdraw_stake(
+                &spl_stake_pool::ID,
+                self.stake_pool.key,
+                self.validator_stake_list.key,
+                self.stake_pool_withdraw_authority.key,
+                self.stake_account_to_split.key,
+                self.user_new_stake_account.key,
+                self.user.key,
+                &self.pool_details.key(),
+                &self.pool_token_vault.key(),
+                self.manager_fee_account.key,
+                self.stake_pool_token_mint.key,
+                self.token_program.key,
+                amount,
```

Comment:
I am not familiar with spl_stake_pools, is this amount denominated in SOL or in the stake pool's token (e.g. bSOL)? If in SOL, where does bSOL come into the picture? If in bSOL, we have to first convert the amount passed into the instruction to bSOL.

See here https://github.com/sunrise-stake/app/blob/e8eb8bb0eb0d8332519e4b97d92865d1aef31bed/programs/sunrise-stake/src/lib.rs#L63

Reply:
You're right. The instruction expects the amount in number of pool_tokens, bSOL in this case. I'll keep that in mind as a todo